docs/contract-updatability.md

```diff
@@ -0,0 +1,357 @@
+
+## Validation Goals
+The contract update validation ensures that:
+
+- Stored data doesn't change its meaning when a contract is updated.
+- Decoding and using stored data does not lead to runtime crashes.
+  - For example, it is invalid to add a field because existing stored data won't have the new field.
+  - Loading the existing data will result in garbage/missing values for such fields.
+  - A static check of the access of the field would be valid, but the interpreter would crash when accessing the field,
+    because the field has a missing/garbage value.
+
+## Updating a Contract
+
+A contract may consist of fields and other declarations such as composite types, functions, constructors, etc.
+When an existing contract is updated, all its inner declarations are also validated.
+
+### Contract Fields
+When a contract is deployed, the fields of the contract are stored in an account's contract storage.
+Changing the fields of a contract only changes the way the program treats the data, but does not change the already
+stored data itself, which could potentially result in runtime inconsistencies as mentioned in the previous section.
+
+See the [Fields](#fields) section for the possible updates that can be done to the fields, and the restrictions
+imposed on changing fields of a contract.
+
+### Nested Declarations
+Contracts can have nested composite type declarations such as structs, resources, interfaces, and enums.
+When a contract is updated, its nested declarations are checked, because:
+ - They can be used as type annotation for the fields of the same contract, directly or indirectly.
+ - Any third-party contract can import the types defined in this contract and use them as type annotations.
+ - Hence, changing the type definition is the same as changing the type annotation of such a field (which is also invalid,
+   as in the [Fields](#fields) section).
+
+Changes that can be done to the nested declarations, and the update restrictions are described in following sections:
+ - [Structs, resources and interface](#structs-resources-and-interfaces)
+ - [Enums](#enums)
+ - [Functions](#functions)
+ - [Constructors](#constructors)
+
+## Fields
+A field may belong to a contract, struct, resource, or interface.
+
+#### Valid Changes:
+- Removing a field is valid
+  ```cadence
+  // Existing contract
+
+  pub contract Foo {
+      pub var a: String
+      pub var b: Int
+  }
+
+
+  // Updated contract
+
+  pub contract Foo {
+      pub var a: String
+  }
+  ```
+  - It leaves data for the removed field unused at the storage, as it is no longer accessible.
+  - However, it does not cause any runtime crashes.
+
+- Changing the order of fields is valid.
+  ```cadence
+  // Existing contract
+
+  pub contract Foo {
+      pub var a: String
+      pub var b: Int
+  }
+
+
+  // Updated contract
+
+  pub contract Foo {
+      pub var b: Int
+      pub var a: String
+  }
+  ```
+
+- Changing the access modifier of a field is valid.
```

Comment:
Maybe add a note below explaining that the validation doesn't ensure that other code depending on this code will stay valid, only that the code and the data it stores itself stays valid.

Reply:
That's a really good idea! 
We could actually add it in general. e.g: under the "goals" section it currently says what are the goals are, but it doesn't say what is **not a goal**.